scripts/src/main/resources/scripts/functions

```diff
@@ -346,47 +346,140 @@ function doDevonCommandAndReturn() {
 }
 
 # $1: the URL to download
-# $2: the optional, explicit filename to save to
-# $3: the optional, target directory to save to
+# $2: the optional, target directory to save to
+# $3: the software to download (e.g. 'java', 'mvn', 'node')
+# $4: the version of the software to download
+# $5: operating system name
+# $6: the optional architecture (e.g. x64 or x86_64)
+# $7: the optional extension (e.g. 'tar.gz' or 'zip')
+# $8: the optional code (magic value to avoid tool specific quirks in general functions)
+# $9: optional filename to save to
 function doDownload() {
   doLicenseAgreement
-  local target_dir
-  local tmp_file
-  if [ -z "${3}" ]
+  local url="${1}"
+  local target_dir="${2}"
+  local software="${3}"
+  local version="${4}"
+  local os="${5}"
+  local arch="${6}"
+  local ext="${7}"
+  local code="${8}"
+  local filename"${9}"
+  if [ -z "${target_dir}" ]
   then
     target_dir="${DEVON_DOWNLOAD_DIR}"
-  else
-    target_dir="${3}"
   fi
-  local filename
-  if [ -n "${2}" ]
+  mkdir -p "${target_dir}"
+  local tmp_file
+  tmp_file=$(mktemp -ut "${filename}.XXXXXX")
+  if [ "${url}" = "" ] || [ "${url}" = "-" ]
   then
-    filename="${2}"
+    if [ -z "${arch}" ]
+    then
+      arch="$(uname -m)"
+    fi
+    local filename_fallback="${software}-${version}"
+    if [ -n "${os}" ]
+    then
+      filename_fallback="${filename_fallback}-${os}"
+    fi
+    if [ "${arch}" != "-" ]
+    then
+      filename_fallback="${filename_fallback}-${arch}"
+    fi
+    filename_fallback="${filename_fallback}.${ext}"
+    local mirrors_dir="${DEVON_IDE_HOME}/mirrors"
+    doDebug "Cloning or updating download mirrors"
+    doGitPullOrClone "${mirrors_dir}" "${DEVON_MIRRORS:-https://github.com/devonfw/ide-mirrors.git}"
+    if [ ! -d "${mirrors_dir}/${software}" ]
+    then
+      doFail "No mirror is configured at ${mirrors_dir}/${software}"
+    fi
+    local mirrors_file="${mirrors_dir}/${software}/${version}"
+    if [ ! -e "${mirrors_file}" ]
+    then
+      doDebug "Mirror file not found for version at ${mirrors_file}"
+      mirrors_file="${mirrors_dir}/${software}/default"
+    fi
+    if [ ! -e "${mirrors_file}" ]
+    then
+      doFail "No mirror is configured at ${mirrors_file}"
```

Comment:
This is a breaking change as people would have to migrate their settings towards the new design, which is not intuitive at all.
So enable this change (which I really appreciate), we need to provide automated migration capabilities and notify the user to properly sync his / her settings after automatic migration.
Do you also updated the ide settings accordingly such that we can merge both PRs together only?

Reply:
Clarified on my own. Looks great. Thanks for the effort.